# Breakpoint Calling Analysis

Breakpoint calling analysis runs the following workflows:

- Destruct
- Lumpy
- Gridss
- Svaba

The analysis takes in tumour and normal bam files as input and generates a set of high confidence breakpoints that can be used for downstream analyses. The high confidence breakpoints are breakpoints that were detected by 2 or more callers out of the 4, which was supported by our benchmarking efforts. 

![](https://lucid.app/publicSegments/view/d999a11c-5490-47f3-8103-0d97c2c22034/image.png)

## Input Data Format:

### Tumour:
Multiple bam files:
- generated by the `alignment` analysis from mondrian, as explained [here](data_formats/qc.md#merged-bam)

### Normal:
A normal Bam file:
- generated by the `alignment` analysis on normal cell bams from mondrian as explained [here](data_formats/qc.md#merged-bam), or
- bam file generated by bulk WGS sequencing


### input json file

Please see [breakpoint input json](data_formats/breakpoint_calling.md#params-yaml)


### Output:

| File                                    | Description                                                                                                     |
|-----------------------------------------|-----------------------------------------------------------------------------------------------------------------|
|four_way_consensus.csv.gz                |consensus between all callers. See [consensus](data_formats/breakpoint_calling.md#consensus)                     |
|metadata.yaml                            |Please see [metadata](data_formats/metadata_yaml_output.md#breakpoint-calling)                                   |
|{sample_id}_breakpoint_library_table.csv |One file per sample. Please see [Destruct Library](data_formats/breakpoint_calling.md#destruct-library)          |
|{sample_id}_breakpoint_read_table.csv    |One file per sample. Please see [Destruct Reads](data_formats/breakpoint_calling.md#destruct-reads)              |
|{sample_id}_breakpoint_table.csv         |One file per sample. Please see [Destruct Table](data_formats/breakpoint_calling.md#destruct-table)              |
|{sample_id}_gridss.vcf.gz                |One file per sample. Please see [Gridss](data_formats/breakpoint_calling.md#gridss-vcf)                          |
|{sample_id}_lumpy.vcf                    |One file per sample. Please see [Lumpy](data_formats/breakpoint_calling.md#lumpy-vcf)                            |
|{sample_id}.svaba.somatic.sv.vcf.gz      |One file per sample. Please see [Svaba](data_formats/breakpoint_calling.md#svaba-vcf)                            |


#### NOTES:
- adapters must be trimmed.
- ideally, no secondary alignments in bam files
