# snv genotyping

Snv genotyping pipeline runs:

1. vartrix
2. pysam based genotyper

The analysis takes in json file as input and generates one csv file for each analysis above. 

![](https://lucid.app/publicSegments/view/cdc0e5ee-a6d3-43f7-b32e-30c74e331d20/image.png)



## Input Data Format:

The input to the pipeline is a json file.

Please see [Snv Genotyping Input Json](data_formats/snv_genotyping.md#input-json) for detailed explanation. 


### Tumour bam file:
- generated by the `alignment` analysis from mondrian, as explained [here](data_formats/alignment.md#merged-bam)

### Snv vcf file:
- must be bgzip'd and indexed with tabix
- pipeline requires chrom, pos, ref and alt columns only



### Output:
| File                  | Description                                                                                                   |
|-----------------------|---------------------------------------------------------------------------------------------------------------|
| metadata.yaml         | Please see [metadata](data_formats/metadata_yaml_output.md#snv-genotyping)                                    |
| snv_genotyping.csv.gz | per cell supporting read counts. Please see [pysam genotyper](data_formats/snv_genotyping.md#pysam-genotyper) |
| vartrix_parsed.csv.gz |per cell supporting read counts. Please see [vartrix](data_formats/snv_genotyping.md#vartrix)                  |
