# Variant calling Analysis

Variant calling analysis runs the following workflows:

- Mutationseq
- Strelka
- Mutect

The analysis takes in one or more tumour files and a single normal bam file as input.  The above tools are used to predict somatic SNVs and InDels, with the output being a set of variant calls in VCF per input tumour BAM.  Multiple tumour bam inputs are provided as a way of allowing for the input of sequencing from multiple scWGS runs, or as a way to split very large datasets into managable sets of bams.  VCFs from each tumour BAM and each variant calling tool are input into a consensus algorithm to identify a set of high confidence variant calls.  Consensus VCFs from each tumour BAM are merged (union set) and converted to MAF using vcf2maf.  The workflow outputs a vcf per tumour bam input, per analysis type, a consensus VCF per tumour BAM, and a single MAF file of high confidence SNVs and InDels.

![](https://lucid.app/publicSegments/view/863a76d0-b7ef-408c-bb08-07597fd9fda8/image.png)

## Input Data Format:

### Tumour:
Multiple bam files:
- generated by the `alignment` analysis from mondrian, as explained [here](data_formats/alignment.md#merged-bam)

### Normal:
A normal Bam file:
- generated by the `alignment` analysis on normal cell bams from mondrian as explained [here](data_formats/alignment.md#merged-bam), or
- bam file generated by bulk WGS sequencing



### input json file

Please see [variant input json](data_formats/variant_calling.md#input-json)


### Output:
| File                               | Description                                                                                                                        |
|------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
|final_maf_all_samples.maf           | Maf file with data from all input samples. Please see [all samples maf](data_formats/variant_calling.md#maf-file-with-all-samples) |
|final_vcf_all_samples.vcf.gz        | Vcf file with data from all input samples. Please see [all samples vcf](data_formats/variant_calling.md#vcf-file-with-all-samples) |
|final_vcf_all_samples.vcf.gz.csi    | csi index file                                                                                                                     |
|final_vcf_all_samples.vcf.gz.tbi    | tbi index file                                                                                                                     |
|metadata.yaml                       | Please see [metadata](data_formats/metadata_yaml_output.md#variant-calling)                                                                      |
|{sampleid}_consensus.vcf.gz         | Vcf file with consensus calls for {sampleid}. Please see [consensus vcf](data_formats/variant_calling.md#consensus-vcf)            |
|{sampleid}_consensus.vcf.gz.csi     | csi index file                                                                                                                     |
|{sampleid}_consensus.vcf.gz.tbi     | tbi index file                                                                                                                     |
|{sampleid}_museq.vcf.gz             | museq vcf file for {sampleid}. Please see [museq vcf](data_formats/variant_calling.md#museq-vcf)                                   |
|{sampleid}_museq.vcf.gz.csi         | csi index file                                                                                                                     |
|{sampleid}_museq.vcf.gz.tbi         | tbi index file                                                                                                                     |
|{sampleid}_mutect.vcf.gz            | mutect vcf file for {sampleid}. Please see [mutect vcf](data_formats/variant_calling.md#mutect-vcf)                                |
|{sampleid}_mutect.vcf.gz.csi        | csi index file                                                                                                                     |
|{sampleid}_mutect.vcf.gz.tbi        | tbi index file                                                                                                                     |
|{sampleid}_strelka_indel.vcf.gz     | strelka vcf file for {sampleid}. Please see [strelka vcf indel](data_formats/variant_calling.md#strelka-vcf)                       |
|{sampleid}_strelka_indel.vcf.gz.csi | csi index file                                                                                                                     |
|{sampleid}_strelka_indel.vcf.gz.tbi | tbi index file                                                                                                                     |
|{sampleid}_strelka_snv.vcf.gz       | strelka vcf file for {sampleid}. Please see [strelka vcf snv](data_formats/variant_calling.md#strelka-vcf)                         | 
|{sampleid}_strelka_snv.vcf.gz.csi   | csi index file                                                                                                                     |
|{sampleid}_strelka_snv.vcf.gz.tbi   | tbi index file                                                                                                                     |
|{sampleid}_updated_counts.maf       | consensus maf file for {sampleid}. Please see [consensus maf](data_formats/variant_calling.md#consensus-maf)                       |

