

## Metadata

The yaml file has the following sections:


1. Files:
this section is a list of files that were generated by the pipeline along with an identifier.

2. Meta:
This section contains metadata about the input data: 



### alignment
```
files:
  alignment_gc_metrics.csv.gz:
    result_type: alignment_gc_metrics
  alignment_metrics.csv.gz:
    result_type: alignment_metrics
  alignment_metrics.csv.gz.yaml:
    result_type: alignment_metrics
  alignment_metrics.tar.gz:
    result_type: alignment_metrics_plots
  all_cells_bulk.bam:
    filtering: passed
    result_type: merged_cells_bam
  all_cells_bulk.bam.bai:
    filtering: passed
    result_type: merged_cells_bam
  all_cells_bulk_contaminated.bam:
    filtering: contaminated
    result_type: merged_cells_bam
  all_cells_bulk_contaminated.bam.bai:
    filtering: contaminated
    result_type: merged_cells_bam
  all_cells_bulk_control.bam:
    filtering: control
    result_type: merged_cells_bam
  all_cells_bulk_control.bam.bai:
    filtering: control
    result_type: merged_cells_bam
  detailed_fastqscreen_breakdown.csv.gz:
    result_type: detailed_fastqscreen_breakdown
  detailed_fastqscreen_breakdown.csv.gz.yaml:
    result_type: detailed_fastqscreen_breakdown
meta:
  cells:
  - SA1090-A96213A-R20-C28
  - SA1090-A96213A-R22-C43
  lanes:
    FL001:
      L001:
        sequencing_centre: BCCRC
  libraries:
  - 128688A
  name: alignment
  samples:
  - SA1090
  version: 0.0.11
```

### hmmcopy
```
files:
  hmmcopy_heatmap.pdf:
    result_type: hmmcopy_heatmap_plots
  hmmcopy_metrics.csv.gz:
    result_type: hmmcopy_metrics
  hmmcopy_metrics.csv.gz.yaml:
    result_type: hmmcopy_metrics
  hmmcopy_params.csv.gz:
    result_type: hmmcopy_params
  hmmcopy_params.csv.gz.yaml:
    result_type: hmmcopy_params
  hmmcopy_reads.csv.gz:
    result_type: hmmcopy_reads
  hmmcopy_reads.csv.gz.yaml:
    result_type: hmmcopy_reads
  hmmcopy_segments.csv.gz:
    result_type: hmmcopy_segments
  hmmcopy_segments.csv.gz.yaml:
    result_type: hmmcopy_segments
  hmmcopy_segments_fail.tar.gz:
    filtering: quality_failed
    result_type: hmmcopy_segment_plots
  hmmcopy_segments_pass.tar.gz:
    filtering: quality_passed
    result_type: hmmcopy_segment_plots
meta:
  cells:
  - SA1090-A96213A-R20-C28
  - SA1090-A96213A-R22-C43
  lanes:
    FL001:
      L001:
        sequencing_centre: BCCRC
  libraries:
  - 128688A
  name: hmmcopy
  samples:
  - SA1090
  version: 0.0.11
```

### breakpoint calling

```
files:
  T2-T-A.svaba.somatic.sv.vcf.gz:
    result_type: svaba_vcf
  T2-T-A_breakpoint_library_table.csv:
    result_type: destruct_library
  T2-T-A_breakpoint_read_table.csv:
    result_type: destruct_reads
  T2-T-A_breakpoint_table.csv:
    result_type: destruct_calls
  T2-T-A_gridss.vcf.gz:
    result_type: gridss_vcf
  T2-T-A_lumpy.vcf:
    result_type: lumpy_vcf
  four_way_consensus.csv.gz:
    result_type: breakpoint_consensus
  four_way_consensus.csv.gz.yaml:
    result_type: breakpoint_consensus
meta:
  T2-T-A:
    cell_ids:
    - SA1090-A96213A-R20-C28
    - SA1090-A96213A-R22-C43
    lanes:
      FL001:
        L001:
          sequencing_centre: BCCRC
    libraries:
    - 128688A
    samples:
    - SA1090
  name: breakpoint_calling
  version: 0.0.11
```

### variant calling
```
files:
  SA123T_consensus.vcf.gz:
    result_type: sample_consensus_vcf
  SA123T_consensus.vcf.gz.csi:
    result_type: sample_consensus_vcf
  SA123T_consensus.vcf.gz.tbi:
    result_type: sample_consensus_vcf
  SA123T_museq.vcf.gz:
    result_type: museq_vcf
  SA123T_museq.vcf.gz.csi:
    result_type: museq_vcf
  SA123T_museq.vcf.gz.tbi:
    result_type: museq_vcf
  SA123T_mutect.vcf.gz:
    result_type: mutect_vcf
  SA123T_mutect.vcf.gz.csi:
    result_type: mutect_vcf
  SA123T_mutect.vcf.gz.tbi:
    result_type: mutect_vcf
  SA123T_strelka_indel.vcf.gz:
    result_type: strelka_indel
  SA123T_strelka_indel.vcf.gz.csi:
    result_type: strelka_indel
  SA123T_strelka_indel.vcf.gz.tbi:
    result_type: strelka_indel
  SA123T_strelka_snv.vcf.gz:
    result_type: strelka_snv
  SA123T_strelka_snv.vcf.gz.csi:
    result_type: strelka_snv
  SA123T_strelka_snv.vcf.gz.tbi:
    result_type: strelka_snv
  SA123T_updated_counts.maf:
    result_type: sample_consensus_maf
  final_maf_all_samples.maf:
    result_type: consensus_maf
  final_vcf_all_samples.vcf.gz:
    result_type: consensus_vcf
  final_vcf_all_samples.vcf.gz.csi:
    result_type: consensus_vcf
  final_vcf_all_samples.vcf.gz.tbi:
    result_type: consensus_vcf
meta:
  SA123T:
    cell_ids:
    - SA1090-A96213A-R20-C28
    - SA1090-A96213A-R22-C43
    lanes:
      FL001:
        L001:
          sequencing_centre: BCCRC
    libraries:
    - 128688A
    samples:
    - SA1090
  name: variant_calling
  version: 0.0.11
```

## snv genotyping
```
files:
  snv_genotyping.csv.gz:
    result_type: snv_genotyping_counts
  snv_genotyping.csv.gz.yaml:
    result_type: snv_genotyping_counts
  vartrix_parsed.csv.gz:
    result_type: vartrix_csv
  vartrix_parsed.csv.gz.yaml:
    result_type: vartrix_csv
meta:
  cell_ids:
  - SA1090-A96213A-R20-C28
  - SA1090-A96213A-R22-C43
  lanes:
    FL001:
      L001:
        sequencing_centre: BCCRC
  libraries:
  - 128688A
  name: snv_genotyping
  samples:
  - SA1090
  version: 0.0.11
```
